code/modules/screentips/atom_extensions.dm

```diff
@@ -0,0 +1,72 @@
+ *
+ * This will be very hot regardless of what is in this loop, and will appear on the profiler.
+ * There is unfortunately nothing we can do about this until skin.dmf supports transparent background
+ * labels.
+ *
+ * This may seem like a lot compared to a dictionary set, but this avoids the expensive operation
+ * of dictionary indexing as it doesn't require hashing and just needs some super cheap
+ * variable accesses
+ */
+/atom/MouseEntered(location, control, params)
+	SHOULD_CALL_PARENT(TRUE)
+	// If someone is connected to us in the queue, then we don't need to requeue
+	if (usr.client.hovered_atom)
+		usr.client.hovered_atom = src
+		return
+	// Holds a hard-reference for a single frame
+	usr.client.hovered_atom = src
+	usr.client.screentip_next = SSscreentips.head
+	SSscreentips.head = usr.client
+
+/// Called when a client mouses over this atom
+/atom/proc/on_mouse_enter(client/client)
+	var/screentip_message = "<span class='big' style='line-height: 0.5'>[MAPTEXT(CENTER(capitalize(name)))]</span>"
+	var/datum/screentip_cache/cache = GLOB.screentips_cache["[type]"]
+	if (cache)
+#ifdef DEBUG
+		screentip_message = "<span class='big' style='line-height: 0.5'>[MAPTEXT(CENTER(capitalize(name) + "(Using Cache)"))]</span>"
+#endif
+		if (ishuman(client.mob) && client.mob.get_active_held_item() == null)
+			client.mob.hud_used.screentip.maptext = "<span valign='top'>[screentip_message][cache.attack_hand]</span>"
+		else
+			client.mob.hud_used.screentip.maptext = "<span valign='top'>[screentip_message]</span>"
+	else
+		client.screentip_context.relevant = ishuman(client.mob)
+		client.screentip_context.user = client.mob
+		client.screentip_context.held_item = client.mob?.get_active_held_item()
+		client.screentip_context.access_context = ""
+		client.screentip_context.left_mouse_context = ""
+		client.screentip_context.tool_icon_context = ""
+		client.screentip_context.shift_left_mouse_context = ""
+		client.screentip_context.ctrl_left_mouse_context = ""
+		client.screentip_context.alt_left_mouse_context = ""
+		client.screentip_context.ctrl_shift_left_mouse_context = ""
+		SEND_SIGNAL(src, COMSIG_ATOM_ADD_CONTEXT, client.screentip_context, client.mob)
+		add_context_self(client.screentip_context, client.mob)
+		if (client.screentip_context.relevant)
+			client.mob.hud_used.screentip.maptext = "<span valign='top'>[screentip_message][client.screentip_context.access_context][client.screentip_context.left_mouse_context][client.screentip_context.ctrl_left_mouse_context][client.screentip_context.shift_left_mouse_context][client.screentip_context.alt_left_mouse_context][client.screentip_context.ctrl_shift_left_mouse_context][client.screentip_context.tool_icon_context]</span>"
+		else
+			client.mob.hud_used.screentip.maptext = "<span valign='top'>[screentip_message]</span>"
+		// Cleanup references for the sake of managing hard-deletes
+		client.screentip_context.user = null
+		client.screentip_context.held_item = null
```

Comment:
would you please do `var/datum/screentip_context/screentip_context` instead of taking it from client for every line?